Migration to Python 3.7+ including boto -> boto3

lod opened the pull request:
Basic testing done, passes all unit tests locally and docker container is created successfully

Ready for high level review

lod commented:
I intend to run black to reformat and hide my uglys

But it will completely obscure what was changed so I've held off for now

lod commented:
These were flagged in the code, but pulling them out so that they don't get lost
Some things to ensure are tested

* Need to check partial transfer behaviour, ensure we clean up
* Ensure we can push without list permissions

bcavagnolo commented on ingester/tests/test_ingester.py:
```diff
@@ -22,66 +25,68 @@ def maker():
     return maker
 
 
+@mock.patch.dict(os.environ, {"AWS_REGION":"us-east-1"})
```
Can you use pytest's `monkeypatch` facility instead? Should be `monkeypatch.setenv("AWS_REGION", "us-east-1")`

lod replied:
Thanks, didn't know about monkeypatch

Fixed in [58ceef1](https://github.com/planetlabs/datalake/pull/78/commits/58ceef10364b8b918c670b84da66bbd9882d3175)

bcavagnolo commented on ingester/tests/test_ingester.py:
```diff
@@ -22,66 +25,68 @@ def maker():
         assert r['metadata'] == metadata
 
 
+@mock.patch.dict(os.environ, {"AWS_REGION":"us-east-1"})
```
...and here too.

lod commented on client/test/conftest.py:
```diff
@@ -69,23 +64,24 @@ def archive(archive_maker):
         else:
+            print("A1", url)
             url = urlparse(url)
+            print("A2", url)
```
Fixed in
[e98f2ae](https://github.com/planetlabs/datalake/pull/78/commits/e98f2aee87e2302a955ac676d46536e3b3fcc476)

lod commented on client/test/test_file.py:
```diff
@@ -42,6 +42,9 @@ def get_randfiles(n):
     files = random_files(2)
+    print(files[0])
+    print(files[1])
+
```
Fixed in
[e98f2ae](https://github.com/planetlabs/datalake/pull/78/commits/e98f2aee87e2302a955ac676d46536e3b3fcc476)

lod commented on client/datalake/archive.py:
```diff
@@ -357,18 +364,36 @@ def _mkdirs(self, path):
-            raise InvalidDatalakePath(msg)
-        return k
+    @staticmethod
+    def split_s3_url(url):
```
Fixed in [6e8bc7d](https://github.com/planetlabs/datalake/pull/78/commits/6e8bc7d6183ed1b3032ec39fc3291ac5de35bb9e)

lod commented on .github/workflows/actions.yml:
```diff
@@ -7,7 +7,7 @@ jobs:
       matrix:
         # 4 jobs total with all permutations
-        python: [2.7, 3.5]
+        python: [2.7, 3.10.1]
```
fixed in [449c16d](https://github.com/planetlabs/datalake/pull/78/commits/449c16d36f35e436455b15609861264b825af0a7)

lod commented:
An update as we approach the end of the week.

I've addressed most of the issues raised. Mostly the easier ones.  
The others will be addressed, soon.

Updated the api project from py2 to py3. Required because it uses the client which is now py3, and the container which is now py3. Had previously not done it because it already used boto3 and I naively assumed all was swell.

Some commits to fix up tests which I missed somehow, and tests that were skipped.

Also fixed up the github unit test automation. Now it covers off all the currently supported python releases and all three of the datalake projects. Also made the tests actually pass.

Next big block of work is testing the client upload flagged special cases. Pushing with very constrained permissions, and multipart upload failures. Will also try and expand the unit tests to cover these.

lod commented on Dockerfile:
```diff
@@ -51,7 +26,6 @@ RUN for d in client ingester api; do \
     cd /opt/$d && \
     python setup.py develop -s /usr/local/bin \
         --egg-path ../../../../../opt/$d/ \
-        -d /usr/local/lib/python2.7/site-packages/ \
```
Fixed in 9e420868502e9d4e3e99e00f7c7fc7d6b26b39f2

setup.py develop is actually depreciated.

So I migrated to pip -e . which does what is required (tested) and avoids the python version path

lod commented:
I did some testing of the failed multipart cleanup behaviour, specifically I did a multipart upload and then killed my network link.

The partial upload is still sitting there. Which makes sense, it's impossible to cleanup when you don't have access to do so.

I suggest we implement a lifecycle rule as described at https://aws.amazon.com/blogs/aws-cloud-financial-management/discovering-and-deleting-incomplete-multipart-uploads-to-lower-amazon-s3-costs/

Govcloud doesn't seem to support the S3 lens for analysis. Partial uploads can be found using
`aws s3api list-multipart-uploads --bucket datalake-staging`

There is currently 1 partial in staging (mine), and 477 in prod.
